certgen/certgen.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package certgen generates (pre-)certificates and (pre-)certificate chains.
+package certgen
+
+import (
+	"crypto"
+	crand "crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"math"
+	"math/big"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509/pkix"
+	"github.com/google/monologue/interval"
+)
+
+var timeNowUTC = func() time.Time {
+	return time.Now().UTC()
+}
+
+// CertificateConfig contains details to be used to populate newly created leaf
+// certificates.
+type CertificateConfig struct {
+	// Required fields
+	//
+	// What these are set to, including the zero values if left unset, is what
+	// will appear in the leaf certificates.
+	SubjectCommonName         string
+	SubjectOrganization       string
+	SubjectOrganizationalUnit string
+	SubjectLocality           string
+	SubjectCountry            string
+	SignatureAlgorithm        x509.SignatureAlgorithm
+
+	// Optional fields
+
+	// Prefix is a prefix that will be used in conjunction with the
+	// SubjectCommonName to create a more specific DNS SAN.
+	Prefix string
+	// NotAfterInterval specifies an interval in which the NotAfter time of a
+	// certificate must fall.
+	NotAfterInterval *interval.Interval
+}
+
+// CA is a Certificate Authority that issues synthetic certificates and
+// certificate chains using its RootCert and RootKey.
+type CA struct {
+	RootCert   *x509.Certificate
+	RootKey    crypto.Signer
```

Comment:
I think the reason that I put it as a crypto.Signer, is that the first thing x509.CreateCertificate does is check that priv is a crypto.Signer, otherwise it returns an error.  I figured that as creating certificates is the thing RootKey is used for, instead of letting RootKey be set to other things that potentially don't implement the crypto.Signer interface and then having them return an error, I might as well codify the crypto.Signer requirement on RootKey here?

Reply:
Presumably there's no real need this to work with DSA keys, as the choice of key type is up to us.